### PR TITLE
[js/webgpu] allow to specify callback for profiling data

### PR DESCRIPTION
**Pull request by fs-eire**
### Description
allow to specify callback for profiling data

*Previous*:
```js
ort.env.webgpu.profilingMode = 'default';  // enable profiling

// profiling data will output to console.
```

*Now*:
```js
ort.env.webgpu.profiling = {
  mode: 'default';  // enable profiling
  ondata: (data) => {
    // .. process the profiling data
  }
};

//for each kernel, "ondata" will be called once. only output to console if ondata is not specified.
```

### CODE REVIEW

File: js/web/lib/wasm/jsep/webgpu/program-manager.ts [hunk 1]
```diff
@@ -103,17 +106,33 @@ export class ProgramManager {
...
+                value => ({dims: value.dims, dataType: tensorDataTypeEnumToString(value.dataType)})),
+            kernelId,
+            kernelType: kernelInfo[0],
+            kernelName: kernelInfo[1],
+            startTime,
+            endTime,
+          });
+        } else {
+          // if no callback is provided, print the profiling message to console
+          let inputShapes = '';
```
Comment (axinging) [inline js/web/lib/wasm/jsep/webgpu/program-manager.ts:124]: Nit: move kernelName here?

File: js/web/lib/wasm/jsep/webgpu/program-manager.ts [hunk 2]
```diff
@@ -103,17 +106,33 @@ export class ProgramManager {
...
-          inputShapes += `input[${i}]: [${value.dims}] | ${tensorDataTypeEnumToString(value.dataType)}, `;
-        });
-        let outputShapes = '';
-        buildArtifact.programInfo.outputs.forEach((value, i) => {
-          outputShapes += `output[${i}]: [${value.dims}] | ${tensorDataTypeEnumToString(value.dataType)}, `;
-        });
-        // eslint-disable-next-line no-console
-        console.log(`[profiling] kernel "${kernelId}|${kernelName}" ${inputShapes}${outputShapes}execution time: ${
-            endTime - startTime} ns`);
+        if (this.backend.env.webgpu.profiling?.ondata) {
```
Comment (axinging) [inline js/web/lib/wasm/jsep/webgpu/program-manager.ts:109]: Default startTime is always 0.  And any use case for profilingTimeBase?
``` 
       if (typeof this.backend.profilingTimeBase === 'undefined') {
          this.backend.profilingTimeBase = startTimeU64;
        }
        const startTime = Number(startTimeU64 - this.backend.profilingTimeBase);
        const endTime = Number(endTimeU64 - this.backend.profilingTimeBase);
```

File: js/common/lib/env.ts [hunk 3]
```diff
@@ -92,11 +92,48 @@ export declare namespace Env {
     async?: boolean;
   }
 
+  export interface WebGpuProfilingDataV1TensorMetadata {
+    dims: readonly number[];
+    dataType: string;
+  }
+  export interface WebGpuProfilingDataV1 {
+    version: 1;
```
Comment (fs-eire) [inline js/common/lib/env.ts:100]: add sessionId: number

**Review by guschmue — DISMISSED** [review]

File: js/web/lib/wasm/jsep/webgpu/program-manager.ts (hunk 2 above)
Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/program-manager.ts:109]: `backend.profilingTimeBase` is designed to log the time of the start time of the first completed kernel (`startTime64`). By using this, the startTime and endTime can be relatively small to fit into int53.

File: js/common/lib/env.ts (hunk 3 above)
Comment (fs-eire) [reply js/common/lib/env.ts:100]: seems cannot retrieve session ID here :(

**Comment by fs-eire** [comment]
Close this one. Use #18732 instead.

The reason why I create a new PR is because somehow the check "license/cla" is not triggered for this PR and seems there is no way to manually trigger it unless create a new PR.

**Pull request closed without merging**